### PR TITLE
Cleanup spg

### LINKED ISSUE #4316: LaunchSPG: heap-use-after-free when closing player slots
In 7fca1097eb98d26846cbfb65b03a7307172df2de, ASan reports a heap-use-after-free while setting many players slots to Closed in the singleplayer launch game screen:
```
=================================================================
==18440==ERROR: AddressSanitizer: heap-use-after-free on address 0x622000a69fa0 at pc 0x5644d703b6e7 bp 0x7ffcb28ffcc0 sp 0x7ffcb28ffcb0
READ of size 8 at 0x622000a69fa0 thread T0
    #0 0x5644d703b6e6 in UI::BaseDropdown::set_value() ../src/ui_basic/dropdown.cc:406
    #1 0x5644d703552f in operator() ../src/ui_basic/dropdown.cc:144
    #2 0x5644d70412f6 in invoke /usr/include/boost/function/function_template.hpp:158
    #3 0x5644d702d45d in boost::function0<void>::operator()() const /usr/include/boost/function/function_template.hpp:763
    #4 0x5644d702d0f4 in boost::signals2::detail::void_type boost::signals2::detail::call_with_tuple_args<boost::signals2::detail::void_type>::m_invoke<boost::function<void ()> >(boost::function<void ()>&, boost::signals2::detail::unsigned_meta_array<>, std::tuple<> const&, boost::enable_if<boost::is_void<boost::function<void ()>::result_type>, void>::type*) const /usr/include/boost/signals2/detail/variadic_slot_invoker.hpp:117
    #5 0x5644d702cddb in boost::signals2::detail::void_type boost::signals2::detail::call_with_tuple_args<boost::signals2::detail::void_type>::operator()<boost::function<void ()>, , 0ul>(boost::function<void ()>&, std::tuple<> const&, mpl_::size_t<0ul>) const /usr/include/boost/signals2/detail/variadic_slot_invoker.hpp:90
    #6 0x5644d702c767 in boost::signals2::detail::void_type boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type>::operator()<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (), boost::function<void ()> >, boost::signals2::mutex> > >(boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (), boost::function<void ()> >, boost::signals2::mutex> > const&) const /usr/include/boost/signals2/detail/variadic_slot_invoker.hpp:133
    #7 0x5644d702b9ba in boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type>, std::_List_iterator<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (), boost::function<void ()> >, boost::signals2::mutex> > >, boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (), boost::function<void ()> >, boost::signals2::mutex> >::dereference() const /usr/include/boost/signals2/detail/slot_call_iterator.hpp:110
    #8 0x5644d702b3bd in boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type>, std::_List_iterator<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (), boost::function<void ()> >, boost::signals2::mutex> > >, boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (), boost::function<void ()> >, boost::signals2::mutex> >::reference boost::iterators::iterator_core_access::dereference<boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type>, std::_List_iterator<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (), boost::function<void ()> >, boost::signals2::mutex> > >, boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (), boost::function<void ()> >, boost::signals2::mutex> > >(boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type>, std::_List_iterator<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (), boost::function<void ()> >, boost::signals2::mutex> > >, boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (), boost::function<void ()> >, boost::signals2::mutex> > const&) /usr/include/boost/iterator/iterator_facade.hpp:550
    #9 0x5644d702ae5f in boost::iterators::detail::iterator_facade_base<boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type>, std::_List_iterator<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (), boost::function<void ()> >, boost::signals2::mutex> > >, boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (), boost::function<void ()> >, boost::signals2::mutex> >, boost::signals2::detail::void_type, boost::iterators::single_pass_traversal_tag, boost::signals2::detail::void_type const&, long, false, false>::operator*() const /usr/include/boost/iterator/iterator_facade.hpp:656
    #10 0x5644d7029e6f in void boost::signals2::optional_last_value<void>::operator()<boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type>, std::_List_iterator<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (), boost::function<void ()> >, boost::signals2::mutex> > >, boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (), boost::function<void ()> >, boost::signals2::mutex> > >(boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type>, std::_List_iterator<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (), boost::function<void ()> >, boost::signals2::mutex> > >, boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (), boost::function<void ()> >, boost::signals2::mutex> >, boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type>, std::_List_iterator<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (), boost::function<void ()> >, boost::signals2::mutex> > >, boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (), boost::function<void ()> >, boost::signals2::mutex> >) const /usr/include/boost/signals2/optional_last_value.hpp:57
    #11 0x5644d7029037 in void boost::signals2::detail::combiner_invoker<void>::operator()<boost::signals2::optional_last_value<void>, boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type>, std::_List_iterator<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (), boost::function<void ()> >, boost::signals2::mutex> > >, boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (), boost::function<void ()> >, boost::signals2::mutex> > >(boost::signals2::optional_last_value<void>&, boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type>, std::_List_iterator<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (), boost::function<void ()> >, boost::signals2::mutex> > >, boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (), boost::function<void ()> >, boost::signals2::mutex> >, boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type>, std::_List_iterator<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (), boost::function<void ()> >, boost::signals2::mutex> > >, boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (), boost::function<void ()> >, boost::signals2::mutex> >) const /usr/include/boost/signals2/detail/result_type_wrapper.hpp:64
    #12 0x5644d7028332 in boost::signals2::detail::signal_impl<void (), boost::signals2::optional_last_value<void>, int, std::less<int>, boost::function<void ()>, boost::function<void (boost::signals2::connection const&)>, boost::signals2::mutex>::operator()() /usr/include/boost/signals2/detail/signal_template.hpp:243
    #13 0x5644d7027b19 in boost::signals2::signal<void (), boost::signals2::optional_last_value<void>, int, std::less<int>, boost::function<void ()>, boost::function<void (boost::signals2::connection const&)>, boost::signals2::mutex>::operator()() /usr/include/boost/signals2/detail/signal_template.hpp:722
    #14 0x5644d706984a in UI::BaseListselect::handle_mousepress(unsigned char, int, int) ../src/ui_basic/listselect.cc:491
    #15 0x5644d7093b35 in UI::Panel::do_mousepress(unsigned char, int, int) ../src/ui_basic/panel.cc:980
    #16 0x5644d7093a8e in UI::Panel::do_mousepress(unsigned char, int, int) ../src/ui_basic/panel.cc:975
    #17 0x5644d7095361 in UI::Panel::ui_mousepress(unsigned char, int, int) ../src/ui_basic/panel.cc:1181
    #18 0x5644d6972fa3 in WLApplication::handle_mousebutton(SDL_Event&, InputCallback const*) ../src/wlapplication.cc:723
    #19 0x5644d6971f26 in WLApplication::handle_input(InputCallback const*) ../src/wlapplication.cc:638
    #20 0x5644d708cb35 in UI::Panel::do_run() ../src/ui_basic/panel.cc:188
    #21 0x5644d699c6e1 in FullscreenMenuBase::MenuTarget UI::Panel::run<FullscreenMenuBase::MenuTarget>() ../src/ui_basic/panel.h:103
    #22 0x5644d698304b in WLApplication::new_game() ../src/wlapplication.cc:1368
    #23 0x5644d6980500 in WLApplication::mainmenu() ../src/wlapplication.cc:1187
    #24 0x5644d6970141 in WLApplication::run() ../src/wlapplication.cc:496
    #25 0x5644d696ace1 in main ../src/main.cc:42
    #26 0x7fdacddae0b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)
    #27 0x5644d696ab7d in _start (/home/benedikt/wl/herd-simulation/widelands+0x49eb7d)

0x622000a69fa0 is located 1696 bytes inside of 5488-byte region [0x622000a69900,0x622000a6ae70)
freed by thread T0 here:
    #0 0x7fdace8afa37 in operator delete(void*) (/lib/x86_64-linux-gnu/libasan.so.6+0xb2a37)
    #1 0x5644d7161d77 in SinglePlayerActivePlayerGroup::~SinglePlayerActivePlayerGroup() ../src/ui_fsmenu/singleplayersetupbox.h:34
    #2 0x5644d7092abf in UI::Panel::check_child_death() ../src/ui_basic/panel.cc:854
    #3 0x5644d7092b20 in UI::Panel::check_child_death() ../src/ui_basic/panel.cc:857
    #4 0x5644d7092b20 in UI::Panel::check_child_death() ../src/ui_basic/panel.cc:857
    #5 0x5644d7092b20 in UI::Panel::check_child_death() ../src/ui_basic/panel.cc:857
    #6 0x5644d7092b20 in UI::Panel::check_child_death() ../src/ui_basic/panel.cc:857
    #7 0x5644d708cc3f in UI::Panel::do_run() ../src/ui_basic/panel.cc:198
    #8 0x5644d699c6e1 in FullscreenMenuBase::MenuTarget UI::Panel::run<FullscreenMenuBase::MenuTarget>() ../src/ui_basic/panel.h:103
    #9 0x5644d698304b in WLApplication::new_game() ../src/wlapplication.cc:1368
    #10 0x5644d6980500 in WLApplication::mainmenu() ../src/wlapplication.cc:1187
    #11 0x5644d6970141 in WLApplication::run() ../src/wlapplication.cc:496
    #12 0x5644d696ace1 in main ../src/main.cc:42
    #13 0x7fdacddae0b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)

previously allocated by thread T0 here:
    #0 0x7fdace8aef17 in operator new(unsigned long) (/lib/x86_64-linux-gnu/libasan.so.6+0xb1f17)
    #1 0x5644d715de3d in SinglePlayerSetupBox::update() ../src/ui_fsmenu/singleplayersetupbox.cc:161
    #2 0x5644d715d4a3 in operator() ../src/ui_fsmenu/singleplayersetupbox.cc:146
    #3 0x5644d715e81a in __invoke_impl<void, SinglePlayerSetupBox::SinglePlayerSetupBox(UI::Panel*, GameSettingsProvider*, uint32_t, uint32_t)::<lambda(const NoteGameSettings&)>&, const NoteGameSettings&> /usr/include/c++/10/bits/invoke.h:60
    #4 0x5644d715e69d in __invoke_r<void, SinglePlayerSetupBox::SinglePlayerSetupBox(UI::Panel*, GameSettingsProvider*, uint32_t, uint32_t)::<lambda(const NoteGameSettings&)>&, const NoteGameSettings&> /usr/include/c++/10/bits/invoke.h:153
    #5 0x5644d715e4eb in _M_invoke /usr/include/c++/10/bits/std_function.h:291
    #6 0x5644d6f1c5d2 in std::function<void (NoteGameSettings const&)>::operator()(NoteGameSettings const&) const /usr/include/c++/10/bits/std_function.h:622
    #7 0x5644d6f19707 in void Notifications::NotificationsManager::publish<NoteGameSettings>(NoteGameSettings const&) ../src/notifications/notifications_impl.h:74
    #8 0x5644d6f17e53 in void Notifications::publish<NoteGameSettings>(NoteGameSettings const&) ../src/notifications/notifications.h:51
    #9 0x5644d71269ec in FullscreenMenuLaunchSPG::enforce_player_names_and_tribes(Widelands::Map&) ../src/ui_fsmenu/launch_spg.cc:121
    #10 0x5644d71262dd in FullscreenMenuLaunchSPG::update() ../src/ui_fsmenu/launch_spg.cc:101
    #11 0x5644d7125c3a in FullscreenMenuLaunchSPG::clicked_select_map() ../src/ui_fsmenu/launch_spg.cc:79
    #12 0x5644d71256fe in FullscreenMenuLaunchSPG::start() ../src/ui_fsmenu/launch_spg.cc:46
    #13 0x5644d708c90f in UI::Panel::do_run() ../src/ui_basic/panel.cc:170
    #14 0x5644d699c6e1 in FullscreenMenuBase::MenuTarget UI::Panel::run<FullscreenMenuBase::MenuTarget>() ../src/ui_basic/panel.h:103
    #15 0x5644d698304b in WLApplication::new_game() ../src/wlapplication.cc:1368
    #16 0x5644d6980500 in WLApplication::mainmenu() ../src/wlapplication.cc:1187
    #17 0x5644d6970141 in WLApplication::run() ../src/wlapplication.cc:496
    #18 0x5644d696ace1 in main ../src/main.cc:42
    #19 0x7fdacddae0b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)

SUMMARY: AddressSanitizer: heap-use-after-free ../src/ui_basic/dropdown.cc:406 in UI::BaseDropdown::set_value()
Shadow bytes around the buggy address:
  0x0c44801453a0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c44801453b0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c44801453c0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c44801453d0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c44801453e0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
=>0x0c44801453f0: fd fd fd fd[fd]fd fd fd fd fd fd fd fd fd fd fd
  0x0c4480145400: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c4480145410: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c4480145420: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c4480145430: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c4480145440: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==18440==ABORTING
```

To reproduce: Press N, choose an 8-player map, set slot 8 to Closed, then close slot 7, then slot 6 etc. In debug builds, changing a slot takes about half a second – to reproduce this issue, be fast and try to close a slot while another one is still being closed.

### PR DESCRIPTION
**Pull request by jmoerschbach**
Fixes #4316

### CODE REVIEW

**Comment by Noordfrees** [comment]
I merged master to trigger the formatting bot and tidy the diff

File: src/ui_fsmenu/singleplayersetupbox.cc [hunk 1]
```diff
@@ -140,26 +140,25 @@ SinglePlayerSetupBox::SinglePlayerSetupBox(UI::Panel* const parent,
             g_style_manager->font_style(UI::FontStyle::kFsGameSetupHeadings)) {
 	add(&title_, Resizing::kAlign, UI::Align::kCenter);
 	add_space(3 * padding);
-	add(&scrollableBox_, Resizing::kAlign, UI::Align::kCenter);
+	add(&scrollableBox_, Resizing::kExpandBoth);
```
Comment (Noordfrees) [inline src/ui_fsmenu/singleplayersetupbox.cc]: I only see this now… but please use under_scores instead of camelCase in variable names.

**Review by Noordfrees — COMMENTED** [review]
Now the launchSPG screen does not respond when changing the own slot number:
shot0000

File: src/ui_fsmenu/singleplayersetupbox.cc (hunk 1 above)
Comment (jmoerschbach) [reply src/ui_fsmenu/singleplayersetupbox.cc]: done

**Comment by jmoerschbach** [comment]
> 
> 
> Now the launchSPG screen does not respond when changing the own slot number:
> shot0000

fixed, sorry... -.-

**Comment by frankystone** [comment]
Just noticed this in multithreading_new, but according to the image above: It is annoying that the two checkboxes 'Peaceful mode' and 'Custom starting positions' are currently above the 'Type of game' dropdown. The checkboxes are also 'hanging in the air' with no visual anchor.

Suggestion:
* Change the String 'Type of game' into something like 'Configure This Game'
* The 'Configure Game' should get a visual box
* Put the two checkboxes in this 'Configure Game' box

Maybe it's a bit late for my suggestion and should not be part of this PR?

See also: #1587

**Comment by jmoerschbach** [comment]
thanks for pointing me to this (quite old) issue :)
I will address it asap on a new branch, so this one here can go in without anymore delay ;)

**Review by Noordfrees — APPROVED** [review]
LGTM :)

**Pull request merged**